packages/dds/tree/src/feature-libraries/default-schema/schemaChecker.ts

```diff
@@ -0,0 +1,153 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { assert, unreachableCase } from "@fluidframework/core-utils/internal";
+import {
+	MapTree,
+	StoredSchemaCollection,
+	TreeFieldStoredSchema,
+	LeafNodeStoredSchema,
+	ObjectNodeStoredSchema,
+	MapNodeStoredSchema,
+	Multiplicity,
+} from "../../core/index.js";
+import { FullSchemaPolicy } from "../modular-schema/index.js";
+import { allowsValue } from "../valueUtilities.js";
+
+export const enum SchemaValidationErrors {
+	NoError,
+	Field_KindNotInSchemaPolicy,
+	Field_IncorrectMultiplicity,
+	Field_NodeTypeNotAllowed,
+	LeafNode_MissingValue,
+	LeafNode_InvalidValue,
+	LeafNode_FieldsNotAllowed,
+	ObjectNode_FieldCountMismatch,
+	ObjectNode_FieldNotInSchema,
+	Node_MissingSchema,
+	UnknownError,
+}
+
+export function isNodeInSchema(
+	node: MapTree,
+	nodeSchemaCollection: StoredSchemaCollection,
+	schemaPolicy: FullSchemaPolicy,
+): SchemaValidationErrors {
+	// Validate the schema declared by the node exists
+	const schema = nodeSchemaCollection.nodeSchema.get(node.type);
+	if (schema === undefined) {
+		return SchemaValidationErrors.Node_MissingSchema;
+	}
+
+	// Validate the node is well formed according to its schema
+
+	if (schema instanceof LeafNodeStoredSchema) {
+		if (node.value === undefined) {
+			return SchemaValidationErrors.LeafNode_MissingValue;
+		}
+		if (node.fields.size !== 0) {
+			return SchemaValidationErrors.LeafNode_FieldsNotAllowed;
+		}
+		if (!allowsValue(schema.leafValue, node.value)) {
+			return SchemaValidationErrors.LeafNode_InvalidValue;
+		}
+	} else if (schema instanceof ObjectNodeStoredSchema) {
+		if (node.fields.size !== schema.objectNodeFields.size) {
+			return SchemaValidationErrors.ObjectNode_FieldCountMismatch;
+		}
+		for (const [fieldKey, field] of node.fields) {
+			const fieldSchema = schema.objectNodeFields.get(fieldKey);
+			if (fieldSchema === undefined) {
+				return SchemaValidationErrors.ObjectNode_FieldNotInSchema;
+			}
+			const fieldInSchemaResult = isFieldInSchema(
+				field,
+				fieldSchema,
+				nodeSchemaCollection,
+				schemaPolicy,
+			);
+			if (fieldInSchemaResult !== SchemaValidationErrors.NoError) {
+				return fieldInSchemaResult;
+			}
+		}
+	} else if (schema instanceof MapNodeStoredSchema) {
+		for (const field of node.fields.values()) {
+			const fieldInSchemaResult = isFieldInSchema(
+				field,
+				schema.mapFields,
+				nodeSchemaCollection,
+				schemaPolicy,
+			);
+			if (fieldInSchemaResult !== SchemaValidationErrors.NoError) {
+				return fieldInSchemaResult;
+			}
+		}
+	} else {
+		assert(false, "Unknown TreeNodeStoredSchema type");
```

Comment:
while we don't currently support assert tagging with it, we do have a "fail" function defined in this package (in utils) for this usecase, and someday will properly make it work with assert tagging, so it can be used for cases like this.

Reply:
I'm not super clear if you were suggesting we use it now or not 😄 . "someday we'll properly make it work with assert tagging, [thus] it can be used for cases like this [today]." or "someday we'll properly make it work with assert tagging, so [at that time] it can be used for cases like this." But I see it's used extensively today so I'll follow suit.